sphinx/source/docs/dev_guide/writing_tests.rst

```diff
@@ -0,0 +1,417 @@
+
+.. _contributor_guide_writing_tests_integration:
+
+Python integration tests
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Bokeh's Python-focused integration tests help make sure that Bokeh's Python code
+works as intended with the TypeScript code of :term:`BokehJS`.
+
+The Python integration tests use `Selenium`_ with `ChromeDriver`_. The test
+scripts are located in :bokeh-tree:`tests/integration`. The folder structure
+resembles the structure of Bokeh's Python models.
+
+Python integration tests use pytest fixtures to handle the webdriver
+configuration and interaction with Selenium. Depending on which context you
+want to test an object in, choose from ``bokeh_model_page``,
+``single_plot_page``, or ``bokeh_server_page``. See
+:bokeh-tree:`bokeh/_testing/plugins/project.py` for more details.
+
+Follow these guidelines when adding or updating Python integration tests:
+
+Keep your code as simple as possible
+    Try to only include things that are essential to your test. Focus your test
+    on one specific functionality. If possible, write several small tests
+    instead of one complex one.
+
+Use the |bokeh.models| API whenever possible
+    Try to use Bokeh's
+    :ref:`low-level bokeh.models interface <userguide_interfaces_models>`
+    instead of the more high-level
+    :ref:`bokeh.plotting interface <userguide_interfaces_plotting>`.
+
+.. _contributor_guide_writing_tests_bokehjs:
+
+Writing JavaScript tests (BokehJS)
+----------------------------------
+
+To maintain the functionality of all :term:`BokehJS` components, Bokeh includes
+various tests written in TypeScript. If all or parts of your changes affect
+BokehJS' JavaScript code, you should add or update the relevant BokehJS tests.
+
+Bokeh's JavaScript tests use a custom testing framework that **requires Google
+Chrome or Chromium**. You need a recent version of one of these browsers
+available on your system to work with these tests.
+
+Like several other testing frameworks such as `Mocha`_ or `Jasmine`_, the
+BokehJS testing framework uses ``describe()`` and ``it()`` functions to set up
+tests. Assertions are made with ``expect()`` and ``assert()``. Use ``assert()``
+when you need to narrow types.
+
+The BokehJS tests are located in :bokeh-tree:`bokehjs/test`. See
+:ref:`contributor_guide_testing_local_javascript` for information on how to run
+them.
+
+General information on contributing to BokehJS is available in
+:ref:`contributor_guide_bokehjs`.
+
+.. _contributor_guide_writing_tests_bokehjs_unit:
+
+BokehJS unit tests
+~~~~~~~~~~~~~~~~~~
+
+The :term:`BokehJS` unit tests help make sure that the individual sections of
+BokehJS function as expected. The unit tests for BokehJS are located in the
+:bokeh-tree:`bokehjs/test/unit/` folder and sub-folders.
+
+Use `Chai "expect" assertion style <Chai_>`_ when writing unit tests for the
+BokehJS testing framework. See the `API documentation of the Chai Assertion
+Library <Chai documentation_>`_ for more details on this style.
```

Comment:
Referring to Chai's documentation may give people a wrong impression regarding bokehjs' assertion framework. Both are quite similar only on the superficial level. If we need to mention Chai at all, then only to say that our framework was inspired by.